pkg/tsdb/cloudwatch/log_actions.go

```diff
@@ -0,0 +1,318 @@
+	"golang.org/x/sync/errgroup"
+)
+
+func (e *CloudWatchExecutor) executeLogActions(ctx context.Context, queryContext *tsdb.TsdbQuery) (*tsdb.Response, error) {
+	resultChan := make(chan *tsdb.QueryResult, len(queryContext.Queries))
+	eg, ectx := errgroup.WithContext(ctx)
+
+	for _, query := range queryContext.Queries {
+		query := query
+
+		eg.Go(func() error {
+			defer func() {
+				if err := recover(); err != nil {
+					plog.Error("Execute Get Metric Data Query Panic", "error", err, "stack", log.Stack(1))
+					if theErr, ok := err.(error); ok {
+						resultChan <- &tsdb.QueryResult{
+							Error: theErr,
+						}
+					}
+				}
+			}()
+
+			dataframe, err := e.executeLogAction(ectx, queryContext, query)
+			if dataframe == nil {
+				return err
+			}
+
+			dataframeEnc, err := dataframe.MarshalArrow()
+			if err != nil {
+				return err
+			}
+
+			resultChan <- &tsdb.QueryResult{RefId: query.RefId, Dataframes: [][]byte{dataframeEnc}}
+			return nil
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	close(resultChan)
+
+	response := &tsdb.Response{
+		Results: make(map[string]*tsdb.QueryResult),
+	}
+
+	for result := range resultChan {
+		response.Results[result.RefId] = result
+	}
+
+	return response, nil
+}
+
+func (e *CloudWatchExecutor) executeLogAction(ctx context.Context, queryContext *tsdb.TsdbQuery, query *tsdb.Query) (*data.Frame, error) {
+	parameters := query.Model
+	subType := query.Model.Get("subtype").MustString()
+
+	defaultRegion := e.DataSource.JsonData.Get("defaultRegion").MustString()
+	region := parameters.Get("region").MustString(defaultRegion)
+	logsClient, er := e.getLogsClient(region)
+
+	if er != nil {
+		return nil, er
+	}
+
+	var data *data.Frame = nil
+	var err error = nil
+
+	switch subType {
+	case "DescribeLogGroups":
+		data, err = e.handleDescribeLogGroups(ctx, logsClient, parameters)
+	case "GetLogGroupFields":
+		data, err = e.handleGetLogGroupFields(ctx, logsClient, parameters, query.RefId)
+	case "StartQuery":
+		data, err = e.handleStartQuery(ctx, logsClient, parameters, queryContext.TimeRange, query.RefId)
+	case "StopQuery":
+		data, err = e.handleStopQuery(ctx, logsClient, parameters)
+	case "GetQueryResults":
+		data, err = e.handleGetQueryResults(ctx, logsClient, parameters, query.RefId)
+	case "GetLogEvents":
+		data, err = e.handleGetLogEvents(ctx, logsClient, parameters)
+	}
+
+	if data == nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (e *CloudWatchExecutor) handleGetLogEvents(ctx context.Context, logsClient cloudwatchlogsiface.CloudWatchLogsAPI, parameters *simplejson.Json) (*data.Frame, error) {
+	queryRequest := &cloudwatchlogs.GetLogEventsInput{
+		Limit:         aws.Int64(parameters.Get("limit").MustInt64(10)),
+		StartFromHead: aws.Bool(parameters.Get("startFromHead").MustBool(false)),
+	}
+
+	logGroupName, err := parameters.Get("logGroupName").String()
+	if err != nil {
+		return nil, fmt.Errorf("Error: Parameter 'logGroupName' is required")
```

Comment:
[Error strings should not be capitalized](https://github.com/golang/go/wiki/CodeReviewComments#error-strings):
```suggestion
		return nil, fmt.Errorf("error: Parameter 'logGroupName' is required")
```